### ISSUE TITLE
Update mermaid.js version

### ISSUE THREAD

**Issue by DDorch** [rich-iannone/DiagrammeR#475]
This issue is in echo with issues #231, #227, #241, #294, #329, #363, #421, #470, #473 and maybe some others...

The need to update the version of mermaid.js embedded in the DiagrammeR package is pressing.

As suggested in #421, updating the file `/DiagrammeR/htmlwidgets/lib/mermaid/dist/mermaid.slim.min.js` in the installation folder of the package with the last version downloaded from <https://cdn.jsdelivr.net/npm/mermaid/dist/mermaid.min.js>.

One way to always have the latest version of DiagrammeR in the package could be to add an `updateMermaid` function in the package:

```r
#'  Update the mermaid library in the package
#' @param version A [character] of the desired version (the latest by default)
#' @return See value returned by [download.file]
updateMermaid <- function(version = "") {
  url <- "https://cdn.jsdelivr.net/npm/mermaid@version/dist/mermaid.min.js"
  if (version != "") {
    stopifnot(grepl("^[[:digit:]]+\\.[[:digit:]]+\\.[[:digit:]]+$", version))
    version <- paste0("@", version)
  }
  url <- gsub("@version", version, url)
  try(
    download.file(url, 
                  system.file("htmlwidgets/lib/mermaid/dist/mermaid.slim.min.js", 
                  package = "DiagrammeR"))
  )
}
```

There are several issues with this solution:

- this url could not be the best one to use to get an up to date and persistent reference for downloading the mermaid.js library
- I don't know if the CRAN authorises to modify the content of the library by itself. But I don't know where to locate the mermaid.js library file elsewhere

@rich-iannone what do you think of that?

**Comment by DanGustafssonTelgeEnergi** [comment]
I have tried this solution on my local installtion in Windows, and it seems to work (for now). 
My original issue was to do "erDiagrams" using mermaid from within R - now that works if I update the js-library according to the instructions above.
Mermaid seems to be actively developed, by many, so some kind of updating scheme would be in order?

**Comment by DDorch** [comment]
Duplicate with #480

**Comment by Seferin** [comment]
Before this running this function with updateMermaid(version="9.3.0") I was stuck with mermaid 0.3.0 and unable to use the mermaid Entity Relationship class of diagram (erDiagram). This solved my problem.

**Comment by DDorch** [comment]
> @DDorch this appears to break the click functionality introduced in #362. Any idea why?

I'm sorry, this needs further investigations to see what has changed in the library that breaks this functionality. 
At least, that explains why the mermaid library has not been updated yet. Unfortunately, I don't have time these next weeks/months to solve this issue.

**Comment by trafficonese** [comment]
> @DDorch this appears to break the click functionality introduced in #362. Any idea why?

I just tried with the newest mermaid verion. The click functionality seems to be fine for the first example in #362 and the second one does not render so clicking is not even possible.

This code does not work with the new version:
   ```
DiagrammeR("graph LR;A(Rounded)-->B[Squared];B-->C{A Decision};
 C-->D[Square One];C-->E[Square Two];
 style A fill:#E5E25F;  style B fill:#87AB51; style C fill:#3C8937;
 style D fill:#23772C;  style E fill:#B6E6E6;"
    )
```
You will just see:
```
Syntax error in graph
mermaid version 9.4.0
```

**Comment by trafficonese** [comment]
you're right, thats a `grViz`. But since the mermaid example isnt even rendering, I cannot test the click functionality.

**Comment by travis-leith** [comment]
On [mermaid.live](https://mermaid.live/edit#pako:eNo9j8sKwjAQRX8lzEqh_YEEhL52itC6M12EZrRBk9Q0QaT0343Vups5nMvMnaCzEoHC1YmhJ_uaZZvaBiNRbtN0l5-bRxAOZcvyuBZTRkrs1KismRk3RWTlTyFHgy37kGolp6dtGSSg0WmhZLwycUMIB9-jRg40jlK4Gwdu5uiFQQqPlVTeOqAXcR8xARG8bV6mA-pdwFUqlYgf67-FS-jw7bJUmt_XbEkz) it renders if you take out all the style stuff.

**Comment by DavZim** [comment]
As I see it, the problem with the newer Mermaid versions is that detecting clicks would be broken.
This might be due to changes in security (eg `securityLevel: 'loose'`, see also [docs](https://mermaid.js.org/syntax/flowchart.html#interaction)).

I have played around with the newest version to create a MWE in shiny to see if interaction works, this is what I came up with:

```
library(shiny)

txt <- r"{graph TD
  A[[Start]] --> B(Preview)
  B --o C{decide}
  B -.-> D
  C ==> D[(End)]

  click A node_click "This is a Tooltip"
  click B node_click "This is a Tooltip"
  click C node_click "This is a Tooltip"
  click D node_click "This is a Tooltip"
}"

css <- "
g.nodes > g.node > rect { fill: red !important; }
g.nodes > g.node > polygon { fill: red !important; }
g.nodes > g.node { cursor: pointer; }
"

start_mermaid_js <- "
import mermaid from 'https://cdn.jsdelivr.net/npm/mermaid@10/dist/mermaid.esm.min.mjs';
mermaid.initialize({ startOnLoad: true, securityLevel: 'loose' });
"

js <- r"{
node_click = function (id) {
  Shiny.setInputValue("mygraph_click", id, {priority: "event"});
};
}"

ui <- fluidPage(
  tags$head(
    tags$style(HTML(css)),
    tags$script(HTML(start_mermaid_js), type = "module"),
    tags$script(HTML(js)),
  ),
  "Here is a Mermaid JS Diagram",
  div(
    class = "html-widget html-widget-output shiny-report-size html-fill-item",
    id = "mygraph",
    pre(
      class = "mermaid",
      txt
    )
  )
)

server <- function(input, output, session) {
  observe({
    cat(sprintf("R detected that in mygraph Node '%s' was clicked\n",
                input$mygraph_click))
  })
}

shinyApp(ui, server)
# When run and clicked on a node, the following would be seen in the R console
#> R detected that in mygraph Node 'A' was clicked
#> R detected that in mygraph Node 'B' was clicked
#> R detected that in mygraph Node 'C' was clicked
#> R detected that in mygraph Node 'D' was clicked
```

There are two caveats: 

1. renderUI does not work with this setup as mermaid is not rendering the chart again.
2. the callback function (node_click) needs to be set for every node and the id (`mygraph`) for the chart is hardcoded. I had hoped that something like the following would work, but it doesnt as the JS code needs to be rerun **after** mermaid renders a graph

```
// NOT WORKING
var nodes = document.querySelectorAll('g.nodes > g.node');
console.log(nodes);
alert('nodes: ' + nodes.length);

nodes.forEach(function(node) {
  node.style.cursor = 'pointer';
  node.addEventListener('click', function(){
    var node_id = node.id;
    console.log("Click on node " + node_id);
  // TODO figure out the id of the chart "mygraph" automatically
   Shiny.setInputValue("mygraph_click", id, {priority: "event"});
  });
});
```

and of course this is not in the DiagrammeR framework. But I hope this might help @rich-iannone or someone else who has enough insider knowledge to implement this in DiagrammeR.

**Comment by zeigerpuppy** [comment]
Have tried using the `updateMermaid` function, specifically `updateMermaid("11.12.0")`

However it does not seem to work for newer options in `sequenceDiagram`

For example:

```{r}
mermaid("
sequenceDiagram
    actor Editor
    Editor->>Package: Install package
    Editor->>Package: Add Refs
    Editor->>Program: Link package
")
```

this renders in the Mermaid live editor with the `actor` replaced with the approprite icon, but just returns text in R(Markdown).

Am I missing something about how the `updateMermaid()` function should work?
I did try reloading the `DiagrammeR` library and restarted the R session.

Trying on Debian, with RStudio 2024.12.1 Build 563, R version 4.5.1


Edit
------

The issue was just an indentation problem.  Looks like mermaid is sensitive to indentation for the `actor` but not for usual nodes.

The updater function does work (as evidenced by version number returned in error), please disregard above

**Comment by DDorch** [comment]
I now have an alternative solution for generating non-interactive Mermaid diagrams by directly querying images from a Mermaid server. There are three caveats: this approach requires an internet connection, a reliable server, and it does not support interactive features such as click actions. However, it fully meets my current needs.

The code is available here: https://forge.inrae.fr/umr-g-eau/fairify/-/blob/main/R/mermaid.R?ref_type=heads

I'm wondering about the opportunity to integrate it in DiagrammeR as an alternative generator always up to date with the last version of mermaid.js. @rich-iannone?

**Comment by rich-iannone** [comment]
This sounds great! I would gladly accept a PR.
